### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by MannikJ**
Hi,

this PR addresses some issues discussed in #2 and #3.

I decided against caching the classes and against putting the class finding logic in the config as well. I think having the service provider implement the DeferrableProvider interface as is, is already enough to reduce the performance impact in production environments. I had previously overlooked this. 
I could verify that when the DeferrableProvider interface is added, the ServiceProvider is neither created nor registered or booted on normal HTTP requests. Also a robust caching mechanism which ensures that the classes are up to date at all times is not that simple to implement. 

I would vote for listing all the API classes explicitly as well. The auto-discovery "feature" does not provide any benefit for the user, but has a high risk of doing harm. It only saves the package maintainer a few minutes ;)

I also removed the region check to be able to use EU_SANDBOX during tests. I still don't really see why this check is needed except because the region db column for the multiseller setup is defined like this:
```
$table->enum('region', SellingPartnerApi::REGIONS);
```
If we changed this, we could probably remove the REGIONS constant  in SellingPartnerApi (`public const REGIONS = ['NA', 'EU', 'FE'];`) completely.

### CODE REVIEW

File: config/spapi.php [hunk 1]
```diff
@@ -1,8 +1,13 @@
 <?php
 
 return [
+    'registration_enabled' => env('SPAPI_REGISTRATION_ENABLED', true),
+
     'installation_type' => 'single',
 
+    'debug' => env('SPAPI_DEBUG', false),
+    'debug_file' => env('SPAPI_DEBUG_FILE', 'php://output'),
```
Comment (jlevers) [inline config/spapi.php:9]: Could you move these new config options to the end of the config file? I'm guessing fewer people will be interacting with them than with the rest of the config options. Also, a comment describing the purpose of the `registration_enabled` option would be great :)

**Comment by jlevers** [comment]
> The auto-discovery "feature" does not provide any benefit for the user, but has a high risk of doing harm.

Out of curiosity, what do you see as the risks of doing harm? I don't mind updating the list of classes occasionally, but I can basically guarantee I'm going to forgot to add new ones sometimes.

Re: the region value check, I'm fine removing it, but I'd like to keep the enum column for the multi-seller setup. If someone using the library wants to modify the migration when they publish it, they're welcome to, but that constraint prevents accidentally adding bad endpoints.

I made some minor comments – if you wouldn't mind addressing those and running `composer lint`, we should be good to go.

**Review by jlevers — CHANGES_REQUESTED** [review]

File: src/SellingPartnerApiServiceProvider.php [hunk 2]
```diff
@@ -56,6 +50,10 @@ public function boot(): void
      */
     public function register(): void
     {
+        if (!config('spapi.registration_enabled', true)) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (MannikJ) [reply src/SellingPartnerApiServiceProvider.php:55]: Well, not essentially to make mocking easier, but rather to create a setup for testing that ensures that there not any outgoing requests made with real credentials. I now see it more as a security switch. On top of this the developer can easily create mocked objects and can be sure that there only mocked requests made. But actually the value of this switch decreased, after I also the binding from singleton to normal. Before, the singleton instances would directly be created and each of them would ensure there is a valid auth token. So there was basically no way to hook in mocked objects, because this was all done automatically before the test method was actually called.

In other words: With singleton binding 'auto-registration' also meant potential accidental api requests before each test although the tested code might even have nothing to do with spapi at all. 
With normal binding the purpose of this switch is reduced, because these requests are only made when the api classes are actually used. This way it is enough to replace/mock the classes in the tests to avoid accidental requests.

But since there is no built-in mocking and mocking these libraries is very rough to do, I think a hard switch like this still provides some value ;)

**Comment by MannikJ** [comment]
> > The auto-discovery "feature" does not provide any benefit for the user, but has a high risk of doing harm.
> 
> Out of curiosity, what do you see as the risks of doing harm? I don't mind updating the list of classes occasionally, but I can basically guarantee I'm going to forgot to add new ones sometimes.

Well harm is maybe a bit too much. What I mean is that what the ClassFinder library is actually not as easy as it looks. It might involve filesystem operations which is potentially very slow - as we noticed. So without a solid caching mechanism it will just be too slow for many usecases. Also a service provider is one of the worst places for slow code. 
A caching mechanism should probably be implemented very similar to Laravel's event discovery. It's also not impossible to do, but it basically just moves the responsibility to pay attention for new classes from the maintainers of this package to its users.  

> Re: the region value check, I'm fine removing it, but I'd like to keep the enum column for the multi-seller setup. If someone using the library wants to modify the migration when they publish it, they're welcome to, but that constraint prevents accidentally adding bad endpoints.

Not yet deleted and will leave as is.

> I made some minor comments – if you wouldn't mind addressing those and running `composer lint`, we should be good to go.

Yes I'll see that I can make the discussed changes within the next days.

**Comment by jlevers** [comment]
I just made the SP API classes listed explicitly. Once those revisions are ready and the conflicts are merged from that update, should be good to go :)

File: config/spapi.php [hunk 3]
```diff
@@ -25,4 +20,15 @@
         // Valid options are NA, EU, FE
         'endpoint' => env('SPAPI_ENDPOINT_REGION', 'NA'),
     ],
+
+    /**
+     * Can be used as a security switch to turn API class registration
+     * completely off for non-production environments.
+     * If enabled, obtaining instances of API classes might cause
+     * unintended API requests using the configured credentials.
```
Comment (jlevers) [inline config/spapi.php:28]: Could you clarify that these would be authorization-related requests to Amazon's OAuth servers, not SP API requests? Just want to make sure people don't think we're making SP API requests in the background with their credentials.

File: src/SellingPartnerApiServiceProvider.php [hunk 4]
```diff
@@ -90,8 +86,8 @@ private function registerSingleSeller(): void
             'region' => config('spapi.single.endpoint'),
         ]);
 
-        foreach ($this->provides() as $cls) {
-            $this->app->bind(
+        foreach (SellingPartnerApi::API_CLASSES as $cls) {
+            $this->app->singleton(
```
Comment (jlevers) [inline src/SellingPartnerApiServiceProvider.php]: I think that in the merge, this got converted back to `singleton`, unless this was an intentional change

**Review by jlevers — COMMENTED** [review]
Couple tiny comments, once those are addressed this'll be ready to merge :)

File: config/spapi.php (hunk 3 above)
Comment (MannikJ) [reply config/spapi.php]: I changed the comment a little, but I also don't want to go too much into the details. If you have an exact wording you'd like to see, please share.

On the other hand and the more I'm thinking about it...
If it's too complicated to explain why one would need this switch, we should probably just leave it out. 

Before changing the binding, it was really a no-go where each tests just potentially fired auth requests directly when the service provider was loaded. But now since this does not happen anymore, I guess it's OK to leave it to the developers to care that there are no requests made during the tests. 

The only thing it solves for me right now, is that I can use production credentials in my local testing setup and just deactivate using them with this switch. Then I just mock the request for normal test runs, and when I want to test with real requests I just activate this switch, remove the mocking parts and run the tests with the real credentials.

But I don't really see that something like this is enough to include this in the library.

Comment (jlevers) [reply config/spapi.php]: Ok, yes, I agree. The use case for this feels small enough that I'm not sure it's worth the added config complexity for the end user. People can figure out their own prod/test credentials setups themselves, I think.

**Review by jlevers — APPROVED** [review]
Awesome, looking good. Thanks!
